### PR TITLE
Fix define link

### PR DESCRIPTION
**Pull request by ngeiswei**
## Description
Attempt to generalize DefineLink so that it can be used for anything not just in the context of the pattern matcher. It is still under the directory `opencog/atoms/bind`, maybe it should be moved elsewhere (like in `opencog/atoms/core`)

* Modify DefineLink class so that the definition body can be anything (not only a ScopeLink)
* Add a static method DefinieLink::get_definition(const Handle& alias) that return the definition body given a definition alias as in (DefineLink alias body)
* Simplify BetaRedexLink. Warning it assumes the definition body is a ScopeLink, I didn't change that, ultimately I guess it should be able to handle non ScopeLink (constants).

I strongly suspect BetaRedexLink and ExecutionOutputLink (or ApplyLink/ApplicationLink as we might soon rename it) are the same thing and should be unified.

## UTest Results
```
100% tests passed, 0 tests failed out of 79
```

### CODE REVIEW

**Comment by ngeiswei** [comment]
@linas if you could have a look at it that would be cool. Otherwise I'll merge it myself in 48h. Thanks.

**Comment by ngeiswei** [comment]
This pull request partially (or wholly) addresses issue #74.

Also, in the case of a function definition, we may want to add support for defining the variables directly in the DefineLike, that is instead of
```atomese
DefineLink
   SchemaNode "f"
   LambdaLink
      VariableNode "X"
      PlusLink
          VariableNode "X"
          NumberNode "1"
```
one would write
```atomese
DefineLink
   VariableNode "X"
   SchemaNode "f"
   PlusLink
      VariableNode "X"
      NumberNode "1"
```
Similarly to Haskell as one may define f using 2 forms
```haskell
f = \x -> x+1
f x = x+1
```
The practical benefit is to save a LambdaLink.

**Comment by linas** [comment]
The BetaRedexLink and the ExecutionOutputLink are definitely NOT the same thing.  See comments on issue #74 to see why.

Note also: there are a bunch of other issues tangled up that need to be solved, again see comments on issue #74 to see what these are.   Until they are resolved, the DefineLink will be kind-of crippled.

File: opencog/atoms/bind/DefineLink.cc [hunk 1]
```diff
@@ -84,4 +78,23 @@ DefineLink::DefineLink(Link &l)
 	init(l.getOutgoingSet());
 }
 
+Handle DefineLink::get_definition(const Handle& alias) {
+	// Get all DefineLinks associated with that alias, beware that it
+	// will also return DefineLink with that alias as definition body.
+    IncomingSet defs = alias->getIncomingSetByType(DEFINE_LINK);
+
+    // Return the first (supposedly unique) definition
```
Comment (linas) [inline opencog/atoms/bind/DefineLink.cc]: you can throw an error if defs.size() > 1

File: opencog/atoms/bind/DefineLink.cc [hunk 2]
```diff
@@ -84,4 +78,23 @@ DefineLink::DefineLink(Link &l)
...
+    for (LinkPtr defl : defs) {
+	    DefineLinkPtr def(DefineLinkCast(defl->getHandle()));
+	    if (def->get_alias() == alias)
```
Comment (linas) [inline opencog/atoms/bind/DefineLink.cc]: it would be really bizarre if this equality check every failed ... it would be a weird internal bug.

**Comment by linas** [comment]
Yes, OK about collapsing DefineLink and LambdaLink.  However: study Haskell less, and prolog and MUMPS more. the atomspace is very unlike haskell, and much more similar to prolog/mumps.

File: opencog/atoms/bind/DefineLink.cc (hunk 1 above)
Comment (ngeiswei) [inline opencog/atoms/bind/DefineLink.cc]: No, consider the following case
```atomese
DefineLink
   ConceptNode "B"
   ConceptNode "A"
DefineLink
   ConceptNode "C"
   ConceptNode "B"
```
although B is used only once as alias for something else, defs.size() would be 2, that is why I need to go through each of them and check if the alias matches.

File: opencog/atoms/bind/DefineLink.cc (hunk 2 above)
Comment (ngeiswei) [inline opencog/atoms/bind/DefineLink.cc]: I don't think so, it would merely raise the exception "Cannot find defined hypergraph for atom %s".

**Comment by ngeiswei** [comment]
OK, I'll take the time to revive my understanding of Prolog (and look into mumps as well, sounds like shumps, maybe it's fun).

File: opencog/atoms/bind/DefineLink.cc (hunk 1 above)
Comment (linas) [inline opencog/atoms/bind/DefineLink.cc]: ah, right!

File: opencog/atoms/bind/DefineLink.cc (hunk 2 above)
Comment (linas) [inline opencog/atoms/bind/DefineLink.cc]: yeah, but how could that ever happen?    There's no way to construct the define without the definition, and there is no way to erase the definition without also erasing the define link.  Right?

Comment (linas) [inline opencog/atoms/bind/DefineLink.cc]: well, nevermind, having this check is harmless, because stranger things have happened before.

Comment (ngeiswei) [inline opencog/atoms/bind/DefineLink.cc]: It's actually necessary for cases I've given above when the alias is being used as definition body by another alias.

**Pull request merged**